### PR TITLE
ResNet use GlobalAveragePooling2D

### PR DESCRIPTION
**Pull request by hgaiser**
This PR is in response to https://github.com/fchollet/keras/pull/7425#issuecomment-317859425 , where the conclusion was to use `GlobalAveragePooling2D` instead of `AveragePooling2D` with a `(7, 7)` kernel (which would only make sense for images of shape `(224, 224, 3)`).

In addition the average pooling layer is moved inside the `if include_top:`, as per the documentation:
```
        pooling: Optional pooling mode for feature extraction
            when `include_top` is `False`.
            - `None` means that the output of the model will be
                the 4D tensor output *of the
                last convolutional layer*.
```

Note that the old behaviour can be used by setting `include_top=False` and `pooling='avg'`. If this change is not correct, then I am curious about the use of the `pooling` parameter. Without this PR and with `include_top=False`, `pool='avg'` and `input_shape=(224, 224, 3)` there would basically be two `GlobalAveragePooling2D` layers after eachother (one through the `AveragePooling2D` with a `(7, 7)` kernel, which is effectively a `GlobalAveragePooling2D` layer given this input).

@fchollet there is a big chance you don't want to change this either, but I at least wanted to try.

### CODE REVIEW

File: keras/applications/resnet50.py [hunk 1]
```diff
@@ -228,10 +227,8 @@ def ResNet50(include_top=True, weights='imagenet',
     x = identity_block(x, 3, [512, 512, 2048], stage=5, block='b')
     x = identity_block(x, 3, [512, 512, 2048], stage=5, block='c')
 
-    x = AveragePooling2D((7, 7), name='avg_pool')(x)
-
     if include_top:
-        x = Flatten()(x)
+        x = GlobalAveragePooling2D()(x)
```
Comment (JihongJu) [inline keras/applications/resnet50.py]: `Flatten` seems still needed here to reshape the 4D tensor to a 2D tensor.

Comment (hgaiser) [inline keras/applications/resnet50.py]: It appears `GlobalAveragePooling2D` already flattens:
```
import keras
inputs = keras.layers.Input((224, 224, 3))
x = keras.layers.GlobalAveragePooling2D()(inputs)
print(x.shape)
```
```
Using TensorFlow backend.
(?, 3)
```

It calls `K.mean(inputs, axis=[1, 2])`, meaning the parameter `keepdims` of `tf.nn.reduce_mean` is falls, meaning the dims are dropped altogether.

According to the `GlobalAveragePooling2D` docs:
```
    # Output shape
        2D tensor with shape:
        `(batch_size, channels)`
```

**Comment by fchollet** [comment]
In the general case, this is correct. However, this would be a breaking change, which is the main reason why we didn't go with the previous PR. As indicated by the line you had to change in the unit tests. Breaking changes are a big deal. Quite a few people use this network for feature extraction. The cost of making this change would be significant.

So I think we should keep the current implementation, even though it has issues, so as not to break anyone's code. We can feel less bad about it by advancing that the PyTorch official implementation has the exact same "issue" (possibly because it was ported from Keras).

**Comment by hgaiser** [comment]
I figured it'd be closed, at least I wanted this PR to exist in case people
are looking for this issue.
